source/simulator/assemblers/stokes.cc

```diff
@@ -0,0 +1,393 @@
+
+        const double JxW = scratch.finite_element_values.JxW(q);
+
+        for (unsigned int i=0; i<stokes_dofs_per_cell; ++i)
+          {
+            data.local_rhs(i) += (density * gravity * scratch.phi_u[i])
+                                 * JxW;
+
+            if (rebuild_stokes_matrix)
+              for (unsigned int j=0; j<stokes_dofs_per_cell; ++j)
+                {
+                  data.local_matrix(i,j) += ( (use_tensor ?
+                                               eta * 2.0 * (scratch.grads_phi_u[i] * stress_strain_director * scratch.grads_phi_u[j])
+                                               :
+                                               eta * 2.0 * (scratch.grads_phi_u[i] * scratch.grads_phi_u[j]))
+                                              // assemble \nabla p as -(p, div v):
+                                              - (pressure_scaling *
+                                                 scratch.div_phi_u[i] * scratch.phi_p[j])
+                                              // assemble the term -div(u) as -(div u, q).
+                                              // Note the negative sign to make this
+                                              // operator adjoint to the grad p term:
+                                              - (pressure_scaling *
+                                                 scratch.phi_p[i] * scratch.div_phi_u[j]))
+                                            * JxW;
+                }
+          }
+      }
+  }
+
+  template <int dim>
+  void
+  StokesAssembler<dim>::
+  local_assemble_stokes_compressible_diffusion (const double                                     /*pressure_scaling*/,
+                                                const bool                                       rebuild_stokes_matrix,
+                                                internal::Assembly::Scratch::StokesSystem<dim>  &scratch,
+                                                internal::Assembly::CopyData::StokesSystem<dim> &data) const
+  {
+    if (!rebuild_stokes_matrix)
+      return;
+
+    const Introspection<dim> &introspection = this->introspection();
+    const FiniteElement<dim> &fe = this->get_fe();
+    const unsigned int stokes_dofs_per_cell = data.local_dof_indices.size();
+    const unsigned int n_q_points    = scratch.finite_element_values.n_quadrature_points;
+
+    for (unsigned int q=0; q<n_q_points; ++q)
+      {
+        for (unsigned int i=0, i_stokes=0; i_stokes<stokes_dofs_per_cell; /*increment at end of loop*/)
+          {
+            if (introspection.is_stokes_component(fe.system_to_component_index(i).first))
+              {
+                scratch.grads_phi_u[i_stokes] = scratch.finite_element_values[introspection.extractors.velocities].symmetric_gradient(i,q);
+                scratch.div_phi_u[i_stokes]   = scratch.finite_element_values[introspection.extractors.velocities].divergence (i, q);
```

Comment:
OK, if you know that it's not a performance problem, then that's fine with me.

It would just be a shame to undo all of the hard work you've done in finding performance bottlenecks!